lagom-cloud-object-storage-example/account-impl/src/main/java/com/lightbend/lagom/account/impl/readside/AccountReportRepository.java

```diff
@@ -0,0 +1,79 @@
+package com.lightbend.lagom.account.impl.readside;
+
+import akka.Done;
+import com.google.common.base.Preconditions;
+import com.lightbend.lagom.account.impl.AccountEvent;
+import com.lightbend.lagom.account.impl.AccountEvent.DepositExecuted;
+import com.lightbend.lagom.account.impl.AccountEvent.WithdrawExecuted;
+import com.lightbend.lagom.javadsl.persistence.Offset;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class AccountReportRepository {
+
+  private final Map<String, Report> reports = new HashMap<>();
+  private final Storage storage;
+
+  @Inject
+  public AccountReportRepository(Storage storage) {
+    this.storage = storage;
+  }
+
+  /**
+   * Handle the post added event.
+   */
+  CompletionStage<Done> handleEvent(AccountEvent evt, Offset offset) {
+
+    // match events per type
+    if (evt instanceof DepositExecuted) {
+      Report report = reports.getOrDefault(evt.getNumber(), Report.newReport(evt.getNumber()));
+      return save(report.newDeposit(evt.getAmount(), evt.getDateTime()));
```

Comment:
I think the logic updating and resending the reports to Storage can be simplified using Java 8's [`Map#merge`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#merge-K-V-java.util.function.BiFunction-)

Reply:
thanks, will check

Reply:
That won't make the code cleaner. `Map#merge` requires a function `(oldReport, newReport) -> Report`.

There is no `remove` neither in our case.

The `save` method takes also care of producing a payload when needed.

Reply:
Thanks for trying. :-)

I just realised the `handleEvent` is not `synchronised` an it's not idempotent. I don't think we should solve neither of those problems but I'd rather state this class has those problems. 

The synchronization issue could be solved if we make the class an inner class on the event processor which is already thread safe.

I think the idempotence could be achieved if the events included the `Account` and then the `Report` didn't do the arythmetic op again and just copied what's on the `event.account.balance`.

Reply:
The fact that is not `synchronised` is not a problem, because there is no tag sharding and therefore only one processor. We can add a note about it, but again, that's a in-memory repo whose goals is to show case Cloud Object Storage integration via Alpakka connector.

Reply:
About events holding the full state of an Entity I would really not do it. We should only propagate the changes, not the entity state. 

Moreover, what we want in the report is the transactions lines, the previous balance and the new balance.